File: server/src/integrations/spyfu.js
import { env } from '../config/env.js';

/* ============================================================
   SPYFU INTEGRATION — competitor keyword & domain analysis.
   Used by module 6 (MARKET RESEARCH).

   Contract: never throws and never blocks the module. When
   credentials are absent or the upstream call fails, it returns
   `live: false` with a clearly-labelled placeholder so the
   module still produces a report. Callers must surface `live`
   so nobody mistakes placeholder data for real data.
   ============================================================ */

const TIMEOUT_MS = 8000;

/** Shape returned in every case, live or not. */
function shell(domain, { live, note, ...rest }) {
  return {
    provider: 'spyfu',
    domain,
    live,
    note,
    fetchedAt: new Date().toISOString(),
    domainStats: null,
    topKeywords: [],
    topCompetitors: [],
    ...rest,
  };
}

async function getJson(url) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(url, { signal: controller.signal });
    if (!res.ok) throw new Error(`SpyFu responded ${res.status}`);
    return await res.json();
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Fetch competitor/keyword intelligence for a domain.
 *
 * @param {string} domain e.g. "ecofly.io"
 * @param {{ countryCode?: string, limit?: number }} [options]
 * @returns {Promise<object>} Always resolves; check `.live`.
 */
export async function fetchDomainIntelligence(domain, options = {}) {
  const { countryCode = 'US', limit = 10 } = options;

  if (!domain) {
    return shell(domain, { live: false, note: 'No domain supplied — nothing to analyse.' });
  }

  if (!env.spyfu.enabled) {
    return shell(domain, {
      live: false,
      note: 'SPYFU_API_ID / SPYFU_SECRET_KEY not configured. Returning placeholder data.',
      ...placeholderFor(domain, limit),
    });
  }

  const auth = `api_id=${encodeURIComponent(env.spyfu.apiId)}&secret_key=${encodeURIComponent(env.spyfu.secretKey)}`;
  const base = env.spyfu.baseUrl;

  try {
    // These three endpoints are the ones module 6 needs. Adjust the paths to
    // match whichever SpyFu plan the account is on.
    const [domainStats, keywords, competitors] = await Promise.all([
      getJson(`${base}/domain_stats_api/v2/getLatestDomainStats?domain=${encodeURIComponent(domain)}&countryCode=${countryCode}&${auth}`),
      getJson(`${base}/serp_api/v2/seo/getMostValuableKeywords?query=${encodeURIComponent(domain)}&pageSize=${limit}&countryCode=${countryCode}&${auth}`),
      getJson(`${base}/competitors_api/v2/seo/getTopCompetitors?domain=${encodeURIComponent(domain)}&countryCode=${countryCode}&${auth}`),
    ]);

    return shell(domain, {
      live: true,
      note: 'Live SpyFu data.',
      domainStats: domainStats?.results?.[0] ?? domainStats ?? null,
      topKeywords: (keywords?.results ?? []).slice(0, limit),
      topCompetitors: (competitors?.results ?? []).slice(0, limit),
    });
  } catch (error) {
    // Degrade rather than fail the whole report.
    return shell(domain, {
      live: false,
      note: `SpyFu request failed (${error.message}). Returning placeholder data.`,
      ...placeholderFor(domain, limit),
    });
  }
}

/** Deterministic stand-in so downstream scoring stays stable without credentials. */
function placeholderFor(domain, limit) {
  const seed = String(domain).length;
  return {
    domainStats: {
      placeholder: true,
      monthlyOrganicClicks: 1000 * seed,
      monthlyPaidClicks: 120 * seed,
      totalOrganicResults: 40 * seed,
    },
    topKeywords: Array.from({ length: Math.min(limit, 5) }, (_, i) => ({
      placeholder: true,
      keyword: `${domain} keyword ${i + 1}`,
      searchVolume: 5000 - i * 700,
      rank: i + 1,
    })),
    topCompetitors: Array.from({ length: Math.min(limit, 3) }, (_, i) => ({
      placeholder: true,
      domain: `competitor-${i + 1}.example.com`,
      overlapScore: 80 - i * 15,
    })),
  };
}

File: server_python/integrations/spyfu.py
"""SpyFu integration — competitor keyword & domain analysis.
Port of server/src/integrations/spyfu.js; used by module 6 (MARKET RESEARCH).

Contract: never raises and never blocks the module. When credentials are
absent or the upstream call fails, it returns `live: False` with a
clearly-labelled placeholder so the module still produces a report.
Callers must surface `live` so nobody mistakes placeholder data for real data.
"""
import os
import json
from datetime import datetime, timezone
from urllib.parse import quote
from urllib.request import urlopen

TIMEOUT_SECONDS = 8
BASE_URL = 'https://www.spyfu.com/apis'


def _enabled():
    return bool(os.getenv('SPYFU_API_ID') and os.getenv('SPYFU_SECRET_KEY'))


def _shell(domain, live, note, **rest):
    out = {
        'provider': 'spyfu',
        'domain': domain,
        'live': live,
        'note': note,
        'fetchedAt': datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'),
        'domainStats': None,
        'topKeywords': [],
        'topCompetitors': [],
    }
    out.update(rest)
    return out


def _get_json(url):
    with urlopen(url, timeout=TIMEOUT_SECONDS) as res:
        if res.status >= 400:
            raise RuntimeError(f'SpyFu responded {res.status}')
        return json.loads(res.read().decode('utf-8'))


def fetch_domain_intelligence(domain, country_code='US', limit=10):
    """Fetch competitor/keyword intelligence for a domain. Always returns;
    check `.live`."""
    if not domain:
        return _shell(domain, False, 'No domain supplied — nothing to analyse.')

    if not _enabled():
        return _shell(
            domain, False,
            'SPYFU_API_ID / SPYFU_SECRET_KEY not configured. Returning placeholder data.',
            **_placeholder_for(domain, limit),
        )

    auth = f'api_id={quote(os.getenv("SPYFU_API_ID", ""))}&secret_key={quote(os.getenv("SPYFU_SECRET_KEY", ""))}'
    d = quote(str(domain))

    try:
        domain_stats = _get_json(
            f'{BASE_URL}/domain_stats_api/v2/getLatestDomainStats?domain={d}&countryCode={country_code}&{auth}')
        keywords = _get_json(
            f'{BASE_URL}/serp_api/v2/seo/getMostValuableKeywords?query={d}&pageSize={limit}&countryCode={country_code}&{auth}')
        competitors = _get_json(
            f'{BASE_URL}/competitors_api/v2/seo/getTopCompetitors?domain={d}&countryCode={country_code}&{auth}')

        stats_results = (domain_stats or {}).get('results') or []
        return _shell(
            domain, True, 'Live SpyFu data.',
            domainStats=stats_results[0] if stats_results else domain_stats,
            topKeywords=((keywords or {}).get('results') or [])[:limit],
            topCompetitors=((competitors or {}).get('results') or [])[:limit],
        )
    except Exception as error:  # Degrade rather than fail the whole report.
        return _shell(
            domain, False,
            f'SpyFu request failed ({error}). Returning placeholder data.',
            **_placeholder_for(domain, limit),
        )


def _placeholder_for(domain, limit):
    """Deterministic stand-in so downstream scoring stays stable without credentials."""
    seed = len(str(domain))
    return {
        'domainStats': {
            'placeholder': True,
            'monthlyOrganicClicks': 1000 * seed,
            'monthlyPaidClicks': 120 * seed,
            'totalOrganicResults': 40 * seed,
        },
        'topKeywords': [
            {
                'placeholder': True,
                'keyword': f'{domain} keyword {i + 1}',
                'searchVolume': 5000 - i * 700,
                'rank': i + 1,
            }
            for i in range(min(limit, 5))
        ],
        'topCompetitors': [
            {
                'placeholder': True,
                'domain': f'competitor-{i + 1}.example.com',
                'overlapScore': 80 - i * 15,
            }
            for i in range(min(limit, 3))
        ],
    }
